### PR TITLE
Add support for x-ms-parameter-grouping

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
Create parameter group structs as specified by the extension.
The grouping of optional args into an 'options' struct was reworked to
build on this implementation.
Added paramgroupinggroup tests and fixed a few tests that were impacted.

### CODE REVIEW

File: src/generator/models.ts [hunk 1]
```diff
@@ -432,10 +427,8 @@ function generateStruct(lang: Language, props?: Property[]): StructDef {
   return st;
 }
 
-function generateOptionalParamsStruct(optionalParam: Parameter): StructDef {
-  // for the optional params struct, the params are smuggled via GroupProperty
-  const params = (<GroupProperty>(<ObjectSchema>optionalParam.schema).properties![0]).originalParameter;
-  const st = new StructDef(optionalParam.schema.language.go!, undefined, params);
```
Comment (catalinaperalta) [inline src/generator/models.ts:438]: Are parameter groups always optional? Since I'm seeing that it's replacing this.

Comment (jhendrixMSFT) [reply src/generator/models.ts:438]: Parameter groups are specified with [x-ms-parameter-grouping](https://github.com/Azure/autorest/tree/master/docs/extensions#x-ms-parameter-grouping) which is optional.

The creation of the optional params struct was reworked to use this same mechanism, see link 258 in transform.ts.

**Comment by jhendrixMSFT** [comment]
At first I was a bit annoyed about the changing of the optional params arg name, however after thinking about it a bit, it's consistent with explicitly defined param groups so I'm ok with it now (let me know if you disagree).

**Comment by catalinaperalta** [comment]
I'm still thinking about it. Not sure I'm where you are about it yet. But I'm also not strongly opposed at the moment. You're saying that you don't mind it because the name corresponds to the param group name?

**Comment by jhendrixMSFT** [comment]
Yes and is consistent with explicitly defined parameter groups.

**Comment by jhendrixMSFT** [comment]
Fixed https://github.com/Azure/autorest.go/issues/348

**Pull request merged**